CORE-10195 k/client: Added support for `describe_topics`

pgellert opened the pull request:
This introduces a new API on the internal kafka client to describe the topic configs of a topic using the DescribeConfigs Kafka API.

This is going to be used later for topic config syncing for Disaster Recovery.

Fixes https://redpandadata.atlassian.net/browse/CORE-10195

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

michael-redpanda committed:
k/client: Added support for describe_topics

Signed-off-by: Michael Boquard <michael@redpanda.com>
(cherry picked from commit 9d358ba4daf5e29088100df2f1bafe5274801e9e)
(cherry picked from commit 12b34cd1e986a938a967d2071eda9202801c356a)

pgellert committed:
k/client/test: test describe_topics

michael-redpanda commented on src/v/kafka/client/client.cc:
```diff
@@ -577,4 +577,57 @@ ss::future<kafka::fetch_response> client::consumer_fetch(
+        co_return ss::coroutine::exception(std::make_exception_ptr(
+          broker_error(controller, error_code::unknown_server_error)));
+    }
+    for (const auto& res : res.data.results) {
+        if (res.error_code != error_code::none) {
+            co_return ss::coroutine::exception(std::make_exception_ptr(
+              topic_error(model::topic(res.resource_name), res.error_code)));
+        }
+    }
```
suggestion: It's possible for one topic to fail due to authz reasons but another topic to succeed.  Instead of throwing an exception here just return the entire response and let the caller handle it

BenPope replied:
> suggestion: It's possible for one topic to fail due to authz reasons but another topic to succeed. Instead of throwing an exception here just return the entire response and let the caller handle it

Moreover, wouldn't any failure result in maxing out the retry loop?

Let's say a request for topics:
If in the first request,  the first fails due to auth, the second succeeds
In subsequent requests, the first fails due to already existing

Only non-successful requests should be retried, or at least, only retriable errors should be handled as a retry.

michael-redpanda replied:
Yeah these are all good points... and an authz failure, imo, shouldn't be retried by the client but by the caller.  Maybe we should have a chat about erorr mitigation

pgellert replied:
I've changed the behaviour now to not retry on any kafka errors in the DescribeConfigsResponse. The callee can handle these errors. This is similar to how franz-go handles kafka errors on the DescribeConfigsResponse: https://github.com/twmb/franz-go/blob/d39445d4fceb7485b0fe88edaa43840c36d7d259/pkg/kgo/client.go#L4072

michael-redpanda commented on src/v/kafka/client/utils.h:
```diff
@@ -56,8 +56,8 @@ auto retry_with_mitigation(
                 }
-                return fut.then(func).handle_exception(
-                  [&eptr](std::exception_ptr ex) mutable {
+                return fut.then([&func]() mutable { return func(); })
+                  .handle_exception([&eptr](std::exception_ptr ex) mutable {
```
@BenPope I think I talked to you about this a ways back and you may have had concerns about my approach

pgellert replied:
I've undone this commit now

michael-redpanda reviewed (commented):
looks good, one suggestion

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--cbd03efe-c763-4af8-a4fa-55c2868a65cd-->
<!--TEST RESULTS-->
<!--562d44b4-14f7-48bd-9651-78b6c73b62f7-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66275">build#66275</a></summary>


| test_class                | test_method                             | test_arguments                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:--------------------------|:----------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest  | test_decommissioning_rebalancing_node   | {"shutdown_decommissioned": true}                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66275#0196f440-8844-4799-81aa-a38972dcc28a | FLAKY         | 20/21    | upstream reliability is '99.5475113122172'. current run reliability is '95.23809523809523'. drift is 4.30942 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 2, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66275#0196f43e-63a8-4bcd-8c5b-2a141b367139 | FAIL          | 0/1      |                                                                                                                                                                       |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66275#0196f43e-63a8-4bcd-8c5b-2a141b367139 | FAIL          | 0/1      |                                                                                                                                                                       |
| RecreateTopicMetadataTest | test_recreated_topic_metadata_are_valid | {"replication_factor": 3}                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66275#0196f440-8844-4799-81aa-a38972dcc28a | FLAKY         | 17/21    | upstream reliability is '97.11538461538461'. current run reliability is '80.95238095238095'. drift is 16.163 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66325">build#66325</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:--------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66325#0196f95c-09ef-4067-96e9-7dd85604ad68 | FLAKY         | 20/21    | upstream reliability is '97.05882352941177'. current run reliability is '95.23809523809523'. drift is 1.82073 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": true}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66325#0196f95c-09ef-4067-96e9-7dd85604ad68 | FLAKY         | 20/21    | upstream reliability is '97.05882352941177'. current run reliability is '95.23809523809523'. drift is 1.82073 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat commented on src/v/kafka/client/utils.h:
```diff
@@ -56,8 +56,8 @@ auto retry_with_mitigation(
                 }
-                return fut.then(func).handle_exception(
-                  [&eptr](std::exception_ptr ex) mutable {
+                return fut.then([&func]() mutable { return func(); })
```
you can probably drop `mutable` here since the capture is a reference which isn't changing even if the thing it is referring to is changing.

BenPope replied:
The copy is required. If `func` has state, and invoking `func` moves the state, but the request fails, the second invocation during retry is not going to do what you want it to do; especially as using something that's moved from may be undefined behaviour.

There are a few examples, I think the main one is produce, but anything with an `iobuf` or `chunked_vector` are not easily copyable, so the func is likely to move its state.

pgellert replied:
I've undone this commit now

BenPope commented on src/v/kafka/client/client.cc:
```diff
@@ -577,4 +577,57 @@ ss::future<kafka::fetch_response> client::consumer_fetch(
+template<typename T>
+std::optional<chunked_vector<T>>
+copy(const std::optional<chunked_vector<T>>& opt_vec) {
+    return opt_vec ? std::make_optional(opt_vec->copy()) : std::nullopt;
```
FYI:
```suggestion
    return opt_vec.transform(&chunked_vector<T>::copy);
```

michael-redpanda replied:
oooooooooooooo

BenPope commented on src/v/kafka/client/client.cc:
```diff
@@ -577,4 +577,57 @@ ss::future<kafka::fetch_response> client::consumer_fetch(
+ss::future<kafka::describe_configs_response> client::do_describe_topics(
+  chunked_vector<model::topic> topics,
+  std::optional<chunked_vector<ss::sstring>> configuration_keys) {
+    auto controller = _controller;
```
If the intention is to avoid using member state that can change across a yield, then this could be a free function that takes the broker as a parameter. I guess the callee should do the lookup each time, so that the retry is useful.

pgellert replied:
> I guess the callee should do the lookup each time, so that the retry is useful.

franz-go seems to issue the request to a random broker or each try, so I think we can follow that pattern as well. (https://github.com/twmb/franz-go/blob/d39445d4fceb7485b0fe88edaa43840c36d7d259/pkg/kgo/client.go#L2479) 

> If the intention is to avoid using member state that can change across a yield, then this could be a free function that takes the broker as a parameter.
I think it wasn't necessary in the first place (since this is a coroutine and `broker` has an `id()` method), and also the code is much simpler if we do the `_brokers.any()` here instead of further up, so I've left this code as is.

BenPope commented on src/v/kafka/client/transport.h:
```diff
@@ -216,6 +216,8 @@ class transport : public net::base_transport {
         } else if constexpr (std::is_same_v<type, sasl_authenticate_request>) {
             return dispatch(std::move(r), api_version(1));
+        } else if constexpr (std::is_same_v<type, describe_configs_request>) {
+            return dispatch(std::move(r), api_version(4));
```
I'm assuming api version negotiation is out of scope for a while?

pgellert replied:
Yes, I think for DR we don't need api version negotiation because the client would only talk to Redpanda brokers (of the same version). But later we might need this if we want to talk to non-redpanda brokers. cc @michael-redpanda

michael-redpanda replied:
yup, first phase here is going to target RP clusters of version X or greater so we can assume API version support.

BenPope commented on src/v/kafka/client/test/describe_topics.cc:
```diff
@@ -0,0 +1,145 @@
+            return conf;
+        }
+    }
+    return std::nullopt;
```
Perhaps:
```suggestion
    auto it = std::ranges::find(
      configs, target, &kafka::describe_configs_resource_result::name);
    return it != configs.end() ? std::optional{*it} : std::nullopt;
```

BenPope commented on src/v/kafka/client/test/describe_topics.cc:
```diff
@@ -0,0 +1,145 @@
+    info("Describe an unknown topic");
+    const auto unknown_tp = model::topic("non-existent");
+    BOOST_REQUIRE_EXCEPTION(
+      client.describe_topics(unknown_tp).get(),
```
I don't think we'd want this to go through the retry loop.

pgellert replied:
I've updated the error handling now, so it no longer does

BenPope reviewed (commented):
I think there should be more coverage of the error path.

pgellert committed:
k/client: document retry move issue

Document the reasoning why `func` needs to be const invokable.

Mutable lambdas or functors with non-const call operators can
unintentionally mutate or move-from internal state, leading to issues on
retry.

To avoid this, we require `func` to be const-invocable. This ensures
that `func` can be safely re-invoked without side effects.

Instead of making `func` non-const invokable, callers should capture all
state by (const) reference or value, and copy explicitly within the
lambda if needed on each retry.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/2a98834cf1e222e79055a7596553af4a6e81f7fb..b265d086130b8ae88960e30c2072d8891cb58db5): address code review:
- fix/undo the lambda copying change + document reasoning with a static_assert
- align describe_configs implementation with franz-go's (issue request to any broker, don't retry any kafka errors from the describe_configs response)
- minor refactor

pgellert commented:
@BenPope:
> I think there should be more coverage of the error path.

We no longer explicitly throw from `do_describe_topics` if those were the errors you meant here. But I am happy to add more tests if you have any specific error scenarios in mind.

BenPope commented:
> @BenPope:
> 
> > I think there should be more coverage of the error path.
> 
> We no longer explicitly throw from `do_describe_topics` if those were the errors you meant here. But I am happy to add more tests if you have any specific error scenarios in mind.

The error handling is gone, so it looks fine now.

michael-redpanda commented on src/v/kafka/client/utils.h:
```diff
@@ -56,6 +56,18 @@ auto retry_with_mitigation(
                           // ignore failed mitigation
                       });
                 }
+
+                static_assert(
+                  std::is_invocable_v<const Func&>,
+                  "Func must have a const operator(), i.e., it must not be a "
+                  "mutable lambda or functor with non-const call operator. "
+                  "This is to prevent problematic usage patterns where state "
+                  "is moved-from on retries. For example, the following is "
+                  "problematic:\n"
+                  "\tauto bad_func = [movable_state]() mutable {\n"
+                  "\t    auto _ = consume(std::move(movable_state));\n"
+                  "\t};");
+
```
🔥

michael-redpanda reviewed (approved):
lgtm
